pkg/sentry/fsimpl/fuse/regular_file.go

```diff
@@ -49,149 +49,77 @@ func (fd *regularFileFD) PRead(ctx context.Context, dst usermem.IOSequence, offs
-	regularFdReadWriterPool.Put(rw)
-}
-
-// read handles and issues the actual FUSE read request.
-// See ReadToBlocks() regarding its purpose.
-func (rw *regularFDReadWriter) read() {
-	// TODO(gvisor.dev/issue/3237): support indirect IO (e.g. caching):
-	// use caching when possible.
-
-	inode := rw.fd.inode()
+	inode := fd.inode()
 
 	// Reading beyond EOF, update file size if outdated.
-	if rw.off+uint64(rw.size) >= atomic.LoadUint64(&inode.size) {
-		if err := inode.reviseAttr(rw.ctx); err != nil {
-			rw.err = err
-			return
+	if uint64(offset)+size > atomic.LoadUint64(&inode.size) {
+		if err := inode.reviseAttr(ctx); err != nil {
+			return 0, err
 		}
 		// If the offset after update is still too large, return error.
-		if rw.off >= atomic.LoadUint64(&inode.size) {
-			rw.err = io.EOF
-			return
+		if uint64(offset) >= atomic.LoadUint64(&inode.size) {
+			return 0, io.EOF
 		}
 	}
 
 	// Truncate the read with updated file size.
 	fileSize := atomic.LoadUint64(&inode.size)
-	if rw.off+uint64(rw.size) > fileSize {
-		// This uint32 conversion will not overflow.
-		// Since rw.off < fileSize and the difference
-		// between must be less than rw.size to make
-		// the if condition true.
-		rw.size = uint32(fileSize - rw.off)
-	}
-
-	// Send the FUSE_READ request and store the data in rw.
-	rw.buf, rw.n, rw.err = rw.fs().ReadInPages(rw.ctx, rw.fd, rw.off, rw.size)
-}
-
-// ReadToBlocks implements safemem.Reader.ReadToBlocks.
-// Due to a deadlock (both the caller of ReadToBlocks and the kernelTask.Block()
-// will try to acquire the same lock), have to separate the rw.read() from the
-// ReadToBlocks() function. Therefore, ReadToBlocks() only handles copying
-// the result into user memory while read() handles the actual reading.
-func (rw *regularFDReadWriter) ReadToBlocks(dsts safemem.BlockSeq) (uint64, error) {
-	if rw.err != nil {
-		return 0, rw.err
+	if uint64(offset)+size > fileSize {
+		size = fileSize - uint64(offset)
 	}
 
-	if dsts.IsEmpty() {
-		return 0, nil
+	buffers, n, err := inode.fs.ReadInPages(ctx, fd, uint64(offset), uint32(size))
+	if err != nil {
+		return 0, err
 	}
 
 	// TODO(gvisor.dev/issue/3237): support indirect IO (e.g. caching),
 	// store the bytes that were read ahead.
 
-	// The actual number of bytes to copy.
-	var size uint32
-	if rw.size < rw.n {
-		// Read more bytes: read ahead.
-		// This is the common case since FUSE will round up the
-		// size to read to a multiple of usermem.PageSize.
-		size = rw.size
-	} else {
-		size = rw.n
+	// Update the number of bytes to copy for short read.
+	if n < uint32(size) {
+		size = uint64(n)
 	}
 
-	// Assume rw.size is less or equal to dsts.NumBytes().
-	if cp, cperr := safemem.CopySeq(dsts, safemem.BlockSeqOf(safemem.BlockFromSafeSlice(rw.buf[:size]))); cperr != nil {
-		return cp, cperr
+	// Copy the bytes read to the dst.
+	// This loop is intended for fragmented reads.
+	// For the majority of reads, this loop only execute once.
+	var copied int64
```

Comment:
Can we use `uint64` and avoid converting `size` to `int64` to do arithmetic?

Reply:
It's due to `DropFirst64(copied)`